TaskWithSmartPointers/MySharedPtr.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+#include <iostream>
+
+template <class T>
+class MySharedPtr
+{
+private:
+	T* m_ptr = nullptr;
+	long* m_referencesCounter = nullptr;
+	friend std::ostream& operator<<(std::ostream& os, MySharedPtr const& m);
+public:
+	T* get()
+	{
+		return m_ptr;
+	}
+
+	long use_count() const noexcept
+	{
+		if (m_referencesCounter != nullptr)
+		{
+			return *m_referencesCounter;
+		}
+		else
+			return 0;
+	}
+
+	bool unique() const noexcept
+	{
+		return (1 == use_count());
+	}
+
+	void reset(T* p)
+	{
+		if (m_referencesCounter != nullptr)
+		{
+			(*m_referencesCounter)--;
+			if (*m_referencesCounter == 0)
+			{
+				delete m_ptr;
+				delete m_referencesCounter;
+			}
+		}
+		m_ptr = p;
+		m_referencesCounter = new long(1);
+	}
+
+	T& operator*()
+	{
+		return *(this->m_ptr);
+	}
+
+	MySharedPtr() // default constructor
+	{
+		m_ptr = nullptr;
```

Comment:
m_ptr and m_referencesCounter's values are already set in members. You can keep it empty in here if you wish.

Reply:
Done